maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+                        ):
+                            geom_finished = True
+
+            geometries_df["geometry"][g_ind] = geometries_disconnected_df["geometry"][
+                g_ind
+            ]
+
+            if geom_finished == True:
+                geometries_disconnected_df = geometries_disconnected_df.drop(g_ind)
+
+            # Okay, that's all we can do for the disconnected geometries at this level.
+            # Go on to filling in the rest of the overlaps by greatest perimeter.
+
+        g_spatial_index = STRtree(
+            geometries_df["geometry"]
+        )  # Build STRtree for the main geometries
+        g_index_by_iloc = dict(
+            (i, list(geometries_df.index)[i]) for i in range(len(geometries_df))
+        )  # Build indexing dictionary
+
+        print("Assigning order", i + 1, "pieces...")
+        for o_ind in progress(
+            overlaps_df_unused_indices, len(overlaps_df_unused_indices)
+        ):
+            this_overlap = overlaps_df["geometry"][o_ind]
+
+            shared_perimeters = []
+
+            possible_geom_integer_indices = [
+                *set(
+                    numpy.ndarray.flatten(
+                        g_spatial_index.query(overlaps_df["geometry"][o_ind])
+                    )
+                )
+            ]
+            possible_geom_indices = [
+                g_index_by_iloc[k] for k in possible_geom_integer_indices
+            ]
+
+            for g_ind in possible_geom_indices:
+                if (g_ind in list(overlaps_df["polygon indices"][o_ind])) and not (
+                    this_overlap.boundary
+                ).intersection(geometries_df["geometry"][g_ind].boundary).is_empty:
+                    shared_perimeters.append(
+                        (
+                            g_ind,
+                            (this_overlap.boundary)
+                            .intersection(geometries_df["geometry"][g_ind].boundary)
+                            .length,
+                        )
+                    )
+
+            # This possibility came up in a previous version, but I hope it will be
+            # obsolete in this version!
+            if len(shared_perimeters) > 0:
+                max_shared_perim = sorted(shared_perimeters, key=lambda tup: tup[1])[-1]
+                poly_to_add_to = max_shared_perim[0]
+                geometries_df["geometry"][poly_to_add_to] = unary_union(
+                    [geometries_df["geometry"][poly_to_add_to], this_overlap]
+                )
+            else:
+                print(
+                    "Couldn't find a polygon to glue a component of intersection ",
+                    o_ind,
+                    " to",
+                )
+
+    reconstructed_df = geometries_df.copy()
+
+    del reconstructed_df["num components orig"]
+    del reconstructed_df["num components refined"]
+
+    return reconstructed_df
+
+
+# Close the gaps:
```

Comment:
Now that we've transferred this code from a Jupyter notebook, we can probably clean up some of these top-level comments that label functions. Any longform documentation about a function should be moved _inside_ the function as a docstring; that way, it'll show up in folks' IDEs and in our autogenerated HTML documentation.